src/kokkos_abstraction.hpp

```diff
@@ -635,6 +635,36 @@ struct SpaceInstance<Kokkos::Cuda> {
 };
 #endif
 
+struct DeviceDeleter {
+  template <typename T>
+  void operator()(T *ptr) {
+    Kokkos::parallel_for(
+        Kokkos::RangePolicy<DevExecSpace>(0, 1),
+        KOKKOS_LAMBDA(const int i) { ptr->~T(); });
```

Comment:
> The destructor I wrote out is only to let the host-side unique_ptr deallocate the device-side object -- is that what you're referring to?

You're calling the destructor for `T` on the device here.

Reply:
Oh, yes, that's what you're saying.

I get why you need to run the constructor on the device, in order to set the virtual function pointers. I'd need to see a pretty compelling case why you need to call the destructor on the device. I wonder if it makes sense to require `std::is_trivial_destructible`.

Reply:
Right, yes, I forgot about this line, thanks for the reminder. 

I don't see a need now to call the destructor on the device. I think if we're going to remove this destructor line, though, we should check that `T` is trivially destructible, which I will add now.